### ISSUE TITLE
BUG: Auditd blocks other processes when the client socket is slow

### ISSUE THREAD

**Issue by jotacarma90** [linux-audit/audit-userspace#273]
# Description

Hello!
Lately I have been working with auditd and the dispatcher to send the logs to another application. I have encountered some strange behavior when delaying the processing of each message sent to the socket. Let me explain in detail:

I have created a python script that connects to the audit socket and receives the corresponding logs. In addition, I have added a configurable sleep in milliseconds after each recv:
```python
from socket import *
from sys import *
from os import *
from stat import *
from os.path import *
import time

def main():
    # Check parameters
    if len(argv) < 2:
        stderr.write("Syntax: {0} <path> <miliseconds to sleep>\n".format(argv[0]))
        exit(1)

    # Read arguments
    path = argv[1]
    t_sleep = argv[2]

    sock = socket(AF_UNIX, SOCK_STREAM)
    sock.connect(path)

    print("Receiving messages:");

    while True:
        data = sock.recv(65536)
        if not data: break
        print(data.decode())
        time.sleep(int(t_sleep)/1000)

if __name__ == "__main__":
    main()

```
I have added some audit rules to generate a lot of events:
```
-w /root -p wa -k access
-w /root -p wa -k system-locale
-w /root -p wa -k time-change
-w /root -p wa -k identity
-w /root -p wa -k MAC-policy
-w /root -p wa -k logins
-w /root -p wa -k session
-w /root -p wa -k perm_mod
-w /root -p wa -k privileged
-w /etc -p wa -k access
-w /etc -p wa -k system-locale
-w /etc -p wa -k time-change
-w /etc -p wa -k identity
-w /etc -p wa -k MAC-policy
-w /etc -p wa -k logins
-w /etc -p wa -k session
-w /etc -p wa -k perm_mod
-w /etc -p wa -k privileged
-w /var/log -p wa -k access
-w /var/log -p wa -k system-locale
-w /var/log -p wa -k time-change
-w /var/log -p wa -k identity
-w /var/log -p wa -k MAC-policy
-w /var/log -p wa -k logins
-w /var/log -p wa -k session
-w /var/log -p wa -k perm_mod
-w /var/log -p wa -k privileged
-w /usr -p wa -k access
-w /usr -p wa -k system-locale
-w /usr -p wa -k time-change
-w /usr -p wa -k identity
-w /usr -p wa -k MAC-policy
-w /usr -p wa -k logins
-w /usr -p wa -k session
-w /usr -p wa -k perm_mod
-w /usr -p wa -k privileged
```
## Testing in audit 3.X
- Configure audit plugin `/etc/audisp/plugins.d/af_test.conf`:
```
active = no
direction = out
path = builtin_af_unix
type = builtin
args = 0640 /tmp/auditsockettest
format = string
```
- Execute script, in this case with 50 ms of sleeping between each `recv`
`python3 /listen.py /tmp/auditsockettest 50`

And then make use of the system, in this case using `yum` to reinstall some packages (the issue is reproduced with different packages):
`yum reinstall -y kernel*`

The problem is that the time it takes for the system to finish the reinstallation increases considerably. I have seen that the yum process is blocking every so often, making the task excessively long.
I have been investigating the `yum` and `auditd` processes with the `strace` tool and have seen that the crashes occur on calls `lsetxattr` and `chown`:
```
12:14:53.795078 lsetxattr("/lib/modules/4.18.0-348.7.1.el8_5.x86_64/kernel/fs/gfs2", "security.selinux", "system_u:object_r:modules_object"..., 38, 0) = 0
12:15:10.789999 ioctl(1, TCGETS, {B38400 opost isig isig icanon echo ...}) = 0
```
As you can see, there are almost 17 seconds between one and the other.

By sufficiently increasing the size of the `backlog_limit` in `/etc/audit/rules.d/audit.rules`, I have observed that the problem is solved. 
I should also add that there seem to be any cpu load to justify the problem.

All this behavior happens for audit 3.0 or higher. Since for lower versions it changes substantially.

**Comment by stevegrubb** [comment]
Yes, this is how it works. Plugins need to dequeue as fast as possible and queue events internally if they cannot be processed. Another option is that I could add a discard option for overflow_action in auditd.conf.

You can check the running auditd metrics with: auditctl --signal cont ; sleep 1 ; cat /var/run/auditd.state

What can happen is auditd's internal queue fills up. Then it can't dequeue from the kernel. Then the kernel's backlog fills up. That triggers the backlog_wait_time to kick in. You can see what the kernel's state is with auditctl -s. You can set the backlog_wait_time to 0 which will cause events to get lost but won't affect other processes.

**Comment by jotacarma90** [comment]
Thanks a lot for all the info guys.
We are going to consider the implementation of an internal buffer to avoid blocking the thread that receives the logs. Any news about this will be much appreciated. 
Thanks and best regards!

**Comment by stevegrubb** [comment]
I also found one issue that may not be helpful, if overflow_action was set to ignore, it was treated as an error. Commit 35a9c05 should make this better. I also updated the text in the auditd-plugins man page to mention what happens. And I bumped up the default q_depth to hopefully avoid some bursty events.

Also, starting with audit-3.0.3, auditd reads multiple records off of netlink at a time in effort to speed up processing of events. So, this might be noticed more by plugins that were doing OK and may now be backlogged a bit more.

Thanks for reporting the issue. Hopefully it's better documented now.
